docs/developer-docs/security/rust-canister-development-security-best-practices.md

```diff
@@ -425,6 +425,29 @@ If a canister traps or panics in `pre_upgrade`, this can lead to permanently blo
 
 - See [Current limitations of the Internet Computer](https://wiki.internetcomputer.org/wiki/Current_limitations_of_the_Internet_Computer), section "Bugs in `pre_upgrade` hooks"
 
+### Reinstantiate global timers during upgrades
+
+#### Security Concern
+
+Global timers are deactivated upon changes to the canister's Wasm module. The [IC specification](https://internetcomputer.org/docs/current/references/ic-interface-spec#timer) states this as follows:
+
+> The global timer is also deactivated upon changes to the canister's Wasm module (calling install_code, uninstall_code methods of the management canister or if the canister runs out of cycles). In particular, the function canister_global_timer won't be scheduled again unless the canister sets the global timer again (using the System API function ic0.global_timer_set).
+
+Upgrade is a mode of `install_code` and hence the global timers are deactivated during an upgrade.
+
+Since global timers are used internally by the Motoko `Timer` mechanism, the same holds true for Motoko Timer. As explained in the [pull request](https://github.com/dfinity/motoko/pull/3542) under "The upgrade story", the global timer gets jettisoned on upgrade, and the timers need to be set up in the post-upgrade hook.
+
+As explained in [this pull request](https://github.com/dfinity/motoko/pull/3542) under "Opting out", the behavior is different when using Motoko and implementing `system func timer`. The `timer` function will be called after an upgrade. In case your canister was using timers for recurring tasks, the `timer` function would likely set the global timer again for a later time. However, the time between invocations of `timer` would not be consistent as the upgrade triggered an "unexpected" call to `timer`.
+
+Using the rust CDK, the reccuring timer is also lost on upgrade as explained in the API documentation of [set_timer_interval](https://docs.rs/ic-cdk/0.6.9/ic_cdk/timer/fn.set_timer_interval.html).
```

Comment:
Any chance we can link to something else than a hard-coded version number?

Reply:
I checked the latest releases and for some reason it didn't have the timer API. I could change it to 0.6.10 since this seems to be the latest patch with the timer API. I'll put a reminder in my calendar in 3 months to check again if latest has the timer API.